ocaml/typing/mode_intf.mli

```diff
@@ -0,0 +1,400 @@
+
+    val many : lr
+
+    val once : lr
+
+    val zap_to_legacy : (allowed * 'r) t -> Const.t
+  end
+
+  module Uniqueness : sig
+    module Const : sig
+      type t =
+        | Unique
+        | Shared
+
+      include Lattice with type t := t
+    end
+
+    type error = Const.t Solver.error
+
+    include Common with module Const := Const and type error := error
+
+    val shared : lr
+
+    val unique : lr
+
+    val zap_to_legacy : ('l * allowed) t -> Const.t
+  end
+
+  module Value : sig
+    module Monadic : sig
+      include Common with type error = [`Uniqueness of Uniqueness.error]
+
+      val check_const : ('l * 'r) t -> Uniqueness.Const.t option
+    end
+
+    module Comonadic : sig
+      include
+        Common
+          with type error =
+            [ `Regionality of Regionality.error
+            | `Linearity of Linearity.error ]
+
+      val check_const :
+        ('l * 'r) t -> Regionality.Const.t option * Linearity.Const.t option
+
+      val linearity : ('l * 'r) t -> ('l * 'r) Linearity.t
+    end
+
+    module Const :
+      Lattice
+        with type t =
+          Regionality.Const.t * Linearity.Const.t * Uniqueness.Const.t
+
+    type error =
+      [ `Regionality of Regionality.error
+      | `Uniqueness of Uniqueness.error
+      | `Linearity of Linearity.error ]
+
+    type 'd t = ('d Monadic.t, 'd Comonadic.t) monadic_comonadic
+
+    include
+      Common
+        with module Const := Const
+         and type error := error
+         and type 'd t := 'd t
+
+    (* some overriding *)
+    val print :
+      ?raw:bool ->
+      ?verbose:bool ->
+      unit ->
+      Format.formatter ->
+      ('l * 'r) t ->
+      unit
+
+    val check_const :
+      ('l * 'r) t ->
+      Regionality.Const.t option
+      * Linearity.Const.t option
+      * Uniqueness.Const.t option
+
+    val regionality : ('l * 'r) t -> ('l * 'r) Regionality.t
+
+    val uniqueness : ('l * 'r) t -> ('l * 'r) Uniqueness.t
+
+    val linearity : ('l * 'r) t -> ('l * 'r) Linearity.t
+
+    val max_with_uniqueness : ('l * 'r) Uniqueness.t -> (disallowed * 'r) t
+
+    val min_with_uniqueness : ('l * 'r) Uniqueness.t -> ('l * disallowed) t
+
+    val min_with_regionality : ('l * 'r) Regionality.t -> ('l * disallowed) t
+
+    val max_with_regionality : ('l * 'r) Regionality.t -> (disallowed * 'r) t
+
+    val min_with_linearity : ('l * 'r) Linearity.t -> ('l * disallowed) t
+
+    val max_with_linearity : ('l * 'r) Linearity.t -> (disallowed * 'r) t
+
+    val set_regionality_min : ('l * 'r) t -> ('l * disallowed) t
+
+    val set_regionality_max : ('l * 'r) t -> (disallowed * 'r) t
+
+    val set_linearity_min : ('l * 'r) t -> ('l * disallowed) t
+
+    val set_linearity_max : ('l * 'r) t -> (disallowed * 'r) t
+
+    val set_uniqueness_min : ('l * 'r) t -> ('l * disallowed) t
+
+    val set_uniqueness_max : ('l * 'r) t -> (disallowed * 'r) t
+
+    val zap_to_legacy : lr -> Const.t
+  end
+
+  module Alloc : sig
```

Comment:
This type is very, very similar to the type of `Value`. Can they be combined?

Regardless, it would be helpful to have a big comment describing what the semantic difference is between `Value` and `Alloc`. When would I want one? When would I want the other?

Reply:
(Added comments describing the difference).

`Value` is the canonical mode for type checking, while `Alloc` is the mode used in arrow types. Currently they only differ in `Locality` vs `Regionality`. Maybe we can make them share some code, but I'm not sure if that would make the code cleaner.